pootle/apps/pootle_store/views.py

```diff
@@ -437,6 +333,181 @@ def source_language(self):
     def directory(self):
         return self.store.parent
 
+    @lru_cache()
+    def get_object(self):
+        return super(PootleUnitJSON, self).get_object()
+
+
+class ProxyTimelineUser(object):
+
+    def __init__(self, submission):
+        self.submission = submission
+
+    @property
+    def display_name(self):
+        return (
+            self.submission["submitter__full_name"].strip()
+            if self.submission["submitter__full_name"].strip()
+            else self.submission["submitter__username"])
+
+    def get_absolute_url(self):
+        return reverse(
+            'pootle-user-profile',
+            args=[self.submission["submitter__username"]])
+
+    def gravatar_url(self, size=80):
+        email_hash = md5(self.submission['submitter__email']).hexdigest()
+        return (
+            'https://secure.gravatar.com/avatar/%s?s=%d&d=mm'
+            % (email_hash, size))
+
+
+class UnitTimelineJSON(PootleUnitJSON):
+
+    model = Unit
+    pk_url_kwarg = "uid"
+
+    template_name = 'editor/units/xhr_timeline.html'
+    timeline_fields = [
+        "type", "old_value", "new_value", "submitter_id", "creation_time",
+        "field", "quality_check_id",
+        "submitter__username", "submitter__email", "submitter__full_name",
+        "suggestion_id", "suggestion__target_f", "suggestion__user__full_name",
+        "suggestion__user__username", "quality_check__name"]
+
+    _language_code = None
+
+    @cached_property
+    def permission_context(self):
+        self.object = self.get_object()
+        return self.project.directory
+
+    @property
+    def project(self):
+        return self.object.store.translation_project.project
+
+    def get_queryset(self):
+        return Unit.objects.get_translatable(self.request.user).select_related(
+            "store__parent",
+            "store__translation_project__project",
+            "store__translation_project__language",
+            "store__translation_project__directory")
```

Comment:
and if I put `__directory` to `project` it will be less for 1 query.
```select_related(
    "store__translation_project__project__directory",
    "store__translation_project__language")
```